src/ansi-c/library/cprover_contracts.c

```diff
@@ -1418,4 +1422,36 @@ __CPROVER_HIDE:;
 #endif
   }
 }
-#endif
+
+/// \brief Implementation of the `obeys_contract` front-end predicate.
+/// \return True iff a function pointer points to the specified contract.
+///
+/// \details If called in an assumption context, translates to an assignment
+/// `function_pointer = contract`. If called in an assertion context, translates
+/// to an `function_pointer == contract`.
+__CPROVER_bool __CPROVER_contracts_obeys_contract(
+  void (**function_pointer)(void),
+  void (*contract)(void),
+  __CPROVER_contracts_write_set_ptr_t set)
+{
+__CPROVER_HIDE:;
+  __CPROVER_assert(
+    (set != 0) &
+      ((set->assume_requires_ctx == 1) | (set->assert_requires_ctx == 1) |
+       (set->assume_ensures_ctx == 1) | (set->assert_ensures_ctx == 1)),
+    "__CPROVER_obeys_contract is used only in requires or ensures clauses");
+  if((set->assume_requires_ctx == 1) | (set->assume_ensures_ctx == 1))
+  {
+    if(*function_pointer != 0)
+    {
+      *function_pointer = contract;
+      return 1;
+    }
+    return 0;
+  }
+  else
+  {
+    return *function_pointer == contract;
+  }
```

Comment:
Why are you not checking `*function_pointer != 0` on this branch?

Reply:
So this is the tricky part:

The check in the first branch is only meant for assumption contexts.

This is to account for cases where users write `__CPROVER_obeys_contract(ptr, contract) || ptr == NULL`, with the `__CPROVER_obeys_pointer` coming first before the NULL check.

Since the '__CPROVER_obeys_contract(ptr, contract)' translates to an assignment in assumption contexts, after the assignment the pointer cannot be NULL anymore which would make the case `ptr == NULL` unreachable.

Guarding the assignment by the NULL check in this branch, the case `ptr == NULL` remains possible.

We could go one step further and have `ptr == NULL` translate to a guarded assignment `if (ptr ==NULL) { ptr = NULL; }` too, to make sure the value set and the nondet bit pattern are in sync.


The other branch is taken in assertion contexts and we don't have to manage this kind of situations, we just check equality.

Reply:
This piece of code really needs documentation. I think it's very hard for anyone other than yourself to see how these pieces fit together, making it impossible to maintain this code.

Reply:
added developer doc in `src/goto-instrument/contracts/doc/`